File: java/AddBinary.java
/*
 * Filename: AddBinary.java
 * Author: Jun Xu
 * Email: jxu.vimacs@gmail.com
 * Created Time: Tue 19 May 2015 02:00:37 PM EDT
 */

public class Solution {
	public String addBinary(String a, String b) {
		if (a == null || a.length() == 0) {
			return b;
		}

		if (b == null || b.length() == 0) {
			return a;
		}

		if (a.length() < b.length()) {
			String tmp = a;
			a = b;
			b = tmp;
		}

		int digit;
		int carry = 0;
		String res = "";

		for (int i = a.length() - 1; i >= a.length() - b.length(); i--) {
			digit = carry + a.charAt(i) - '0' + b.charAt(i - a.length() + b.length()) - '0';
			carry = digit / 2;
			digit %= 2;
			res = digit + res;
		}

		for (int i = a.length() - b.length() - 1; i >= 0; i--) {
			digit = carry + a.charAt(i) - '0';
			carry = digit / 2;
			digit %= 2;
			res = digit + res;
		}

		if (carry != 0) {
			res = carry + res;
		}

		return res;
	}
}

File: cpp/addBinary.cpp
/*
 * Filename: addBinary.cpp
 * Author: Jun Xu
 * Email: junx@cs.bu.edu
 * Created Time: Wed 27 Jan 2016 08:45:55 PM EST
 * Description: 
 */

#include <iostream>

using namespace std;

class Solution
{
	public:
		string addBinary(string a, string b)
		{
			int carry = 0;
			int i = a.length() - 1;
			int j = b.length() - 1;
			string res = "";

			while (i >= 0 && j >= 0)
			{
				carry += (a[i] - '0') + (b[j] - '0');
				res = to_string(carry % 2) + res;
				carry /= 2;
				i--;
				j--;
			}

			while (i >= 0)
			{
				carry += a[i] - '0';
				res = to_string(carry % 2) + res;
				carry /= 2;
				i--;
			}

			while (j >= 0)
			{
				carry += b[j] - '0';
				res = to_string(carry % 2) + res;
				carry /= 2;
				j--;
			}

			if (carry)
			{
				res = to_string(carry) + res;
			}

			return res;
		}
};
